ros/angel_system_nodes/angel_system_nodes/hmm.py

```diff
@@ -0,0 +1,254 @@
+        # TODO: Fill in time remaining?
+
+        self._publisher.publish(message)
+
+        # Update state tracking vars
+        self._previous_step = self._current_step
+
+    def query_task_graph_callback(self, request, response):
+        """
+        Populate the `QueryTaskGraph` response with the task list
+        and return it.
+        """
+        log = self.get_logger()
+        task_g = TaskGraph()
+
+        task_g.task_steps = self._hmm.model.class_str
+
+        # TODO: support different task levels?
+        task_g.task_levels = [0] * len(self._hmm.model.class_str)
+
+        response.task_graph = task_g
+        return response
+
+    def thread_run_hmm_state(self):
+        """
+        HMM runtime thread that gets the current state.
+        Thread is awakened each time a new ActivityDetection message is
+        received.
+        """
+        log = self.get_logger()
+        log.info("HMM state thread started")
+
+        while self._hmm_active.wait(0): # will quickly return false if cleared
+            if self._hmm_state_awake_evt.wait(self._hmm_active_heartbeat):
+                log.info("HMM state loop awakened")
+                self._hmm_state_awake_evt.clear()
+
+                # Get the HMM prediction
+                start_time = time.time()
+                self._current_step = self._hmm.get_current_state()
+                log.info(f"HMM get_current_state time: {time.time() - start_time}")
+                log.info(f"Current step {self._current_step}")
+
+                # Publish a new TaskUpdate message
+                self.publish_task_state_message()
+
+    def thread_run_hmm_skip(self):
+        """
+        HMM runtime thread that gets the current likelihood of a step skip.
+        Thread is awakened each time a new ActivityDetection message is
+        received.
+        """
+        log = self.get_logger()
+        log.info("HMM skip thread started")
+
+        while self._hmm_active.wait(0): # will quickly return false if cleared
+            if self._hmm_skip_awake_evt.wait(self._hmm_active_heartbeat):
+                log.info("HMM skip loop awakened")
+                self._hmm_skip_awake_evt.clear()
+
+                # Get the HMM prediction
+                start_time = time.time()
+                skip_score = self._hmm.get_skip_score()
+                log.info(f"Skip score: {skip_score}")
+                log.info(f"HMM get_skip_score time: {time.time() - start_time}")
+
+                # TODO: need to figure out what to do with this score
+
+    def hmm_alive(self) -> bool:
+        """
+        Check that the HMM runtime is still alive and raise an exception
+        if it is not.
+        """
+        alive = (
+            self._hmm_state_thread.is_alive() and self._hmm_skip_thread.is_alive()
+        )
+        if not alive:
+            self.get_logger().warn("HMM thread no longer alive.")
+            self._hmm_thread.join()
```

Comment:
If you do want to stick with both runtime threads, should both be stopped if one stops? If so, then this conditional should likely clear both event variables and then join on both threads before continuing.